doc/design/secret-provisioning.md

```diff
@@ -0,0 +1,167 @@
+# Flexible Secret provisioning with Mystikos
+
+For confidential computing applications, we don't recommend that the
+application code to contain any secrets, as the code (either in source
+or binary forms) could be exposed to hostile environment during
+downloading, uploading, copying, or launching of the application.
+Instead, we recommend the secrets to be securely released to the
+application while it is executed in a Trused Execution Environment (TEE).
+The providers of the secrets could be a key vault service such as AKV,
+M-HSM, or a service run by the user in a locked down and trusted
+environment.
+
+This design document describes the process of secret provisioning
+ with Mystikos in 5 aspects:
+
+1. **the user interface**: how does Mystikos allow users to specify the
+secrets they want in their applications?
+1. **the client library**: how does Mystikos provide a reference secret
+provisioning client library and how can users plug in an alternative library.
+1. **the server library**: how does Mystikos provide a reference secret
+provisioning server library and how can users plug in an alternative library.
+1. **the secret release service**: how does the secret holder use the
+server library for secure release of secrets.
+1. **the attestation model**: the evidences and the verification of
+such evidences agreed by both the client and the server.
+
+## The user interface
+
+Users who wish to provision secrets to their applications running with
+Mystikos can add the following section to config.json:
+
+```json
+secrets: [
+    # Secret 1
+    {
+        "Name": <name of the secret>,
+        "AttestationModel": <a number between 1 to 3. See attestation models below>
+        "HolderURL": <URL to the secret release service>,
+        "Kid": <identifier of the key>,
+        "LocalPath": <Path to the file that will store the secret after its retrieval>,
+        "ClientLibPath": <Path to the client library>
+    },
+    # Secret 2
+    {
+        "Name": <name of the secret>,
+        "AttestationModel": <a number between 1 to 3. See attestation models below>
+        "HolderURL": <URL to the secret release service>,
+        "Kid": <identifier of the key>,
+        "LocalPath": <Path to the file that will store the secret after its retrieval>,
+        "ClientLibPath": <Path to the client library>
+    },
+    ...
+]
+```
+
+Users are responsible for getting the dependent client library into the rootfs consumed
+by Mystikos. They are also responsible for linking with the corresponding server library
+in their services that release the secrets.
+
+## Attestation models
+
+At present, we consider 3 attestation models, from 1 to 3, that can be adopted by the
```

Comment:
Is it necessary to expose the attestation model difference to the app developer? Can the client library abstract out the attestation protocol details?

Reply:
My current design is to have Mystikos runtime making different calls to the client library depending on the attestation model specified in config.json. When a client library supports multiple models, we want the Mystikos runtime to be able to differentiate between the models.

Reply:
For a plugin that supports multiple protocols or options, an alternative to the "predefined attestation model" design is to support an opaque plugin-defined parameter, where the plugin exposes a single "generate_xyz" API, with an optional parameter, and the Mystikos runtime invokes the API with the parameter from the json configuration.

Reply:
The C language is notorious for its difficulty in supporting function overloading and optional parameters. Mystikos is implemented in C, and I think the current design is easier to implement.

Reply:
Do the three "generate_xyz" APIs have different input parameter definition? I was assuming they don't take any input.

Reply:
They take pointers to the buffers that will hold the generated artifacts. They could be augmented with using nonce as one input. But so far no verifier I know of takes advantage of nonces. 

If we convert the attestation models as a parameter to the `generate_xyz` API, I think it actually exposes deeper internals of design and is harder for users to understand, and harder to adapt to a possible future re-implementation.